scheduler/pkg/store/memory.go

```diff
@@ -310,83 +311,65 @@ func (m *MemoryStore) updateLoadedModelsImpl(
 		if !existingState.State.AlreadyLoadingOrLoaded() {
 			logger.Debugf(
 				"Setting model %s version %d on server %s replica %d to LoadRequested",
-				modelKey, modelVersion.version, serverKey, replica.GetReplicaIdx(),
-			)
-			modelVersion.SetReplicaState(replica.GetReplicaIdx(), LoadRequested, "")
-			m.updateReservedMemory(LoadRequested, serverKey, replica.GetReplicaIdx(), modelVersion.GetRequiredMemory())
-			updated = true
-		} else {
-			logger.Debugf(
-				"model %s on server %s replica %d already loaded",
-				modelKey, serverKey, replica.GetReplicaIdx(),
+				modelKey, modelVersion.version, serverKey, replicaIdx,
 			)
+			modelVersion.SetReplicaState(replicaIdx, LoadRequested, "")
+			m.updateReservedMemory(LoadRequested, serverKey, replicaIdx, modelVersion.GetRequiredMemory())
+			replicaStateUpdated = true
 		}
-		updatedReplicas[replica.GetReplicaIdx()] = true
 	}
-	// Unload any existing model replicas assignments no longer needed
-	for replicaIdx, existingState := range modelVersion.ReplicaState() {
-		logger.Debugf(
-			"Looking at replicaidx %d with state %s but ignoring processed %v",
-			replicaIdx, existingState.State.String(), updatedReplicas,
-		)
 
-		if _, ok := updatedReplicas[replicaIdx]; !ok {
-			if !existingState.State.UnloadingOrUnloaded() {
-				if existingState.State == Draining {
-					logger.Debugf(
-						"model %s version %d on server %s replica %d is Draining",
-						modelKey, modelVersion.version, serverKey, replicaIdx,
-					)
-				} else {
-					logger.Debugf(
-						"Setting model %s version %d on server %s replica %d to UnloadEnvoyRequested",
-						modelKey, modelVersion.version, serverKey, replicaIdx,
-					)
-					modelVersion.SetReplicaState(replicaIdx, UnloadEnvoyRequested, "")
-					updated = true
-				}
-			} else {
+	// Unload any existing model replicas assignments that are no longer part of the replica set
+	for replicaIdx, existingState := range modelVersion.ReplicaState() {
+		if _, ok := assignedReplicaIds[replicaIdx]; !ok {
+			if !existingState.State.UnloadingOrUnloaded() && existingState.State != Draining {
 				logger.Debugf(
-					"model %s on server %s replica %d already unloading or can't be unloaded",
-					modelKey, serverKey, replicaIdx,
+					"Setting model %s version %d on server %s replica %d to UnloadEnvoyRequested",
+					modelKey, modelVersion.version, serverKey, replicaIdx,
 				)
+				modelVersion.SetReplicaState(replicaIdx, UnloadEnvoyRequested, "")
+				replicaStateUpdated = true
 			}
 		}
 	}
 
 	// in cases where we did have a previous ScheduleFailed, we need to reflect the change here
 	// this could be in the cases where we are scaling down a model and the new replica count can be all deployed
-	if updated || modelVersion.state.State == ScheduleFailed {
+	// and always send an update for deleted models, so the operator will remove them from k8s
+	if replicaStateUpdated || modelVersion.state.State == ScheduleFailed || model.IsDeleted() {
```

Comment:
Having another thought about the conditions here, I feel that, while the aim is to be specific about which cases are legit to trigger an update, these conditions are a bit arbitrary. In fact there is an issue that I am currently looking at that is going to add yet another condition to this filter. 

This made me realise that if we are already in this function, there is a high probability that there is a status change that triggered it and therefore it might feel that we should try to update the status of the model and send an event downstream. Have you spotted any cases where we should **not** be sending an event?